src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala

```diff
@@ -178,16 +180,60 @@ object ZipAndJarSourcePathFactory extends ZipAndJarFileLookupFactory {
 
 final class FileBasedCache[T] {
   import java.nio.file.Path
-  private case class Stamp(lastModified: FileTime, fileKey: Object)
+  private case class Stamp(size: Long, signatures: ArrayBuffer[Long], fileKey: Object)
   private val cache = collection.mutable.Map.empty[Seq[Path], (Seq[Stamp], T)]
 
+  /**
+    * Read the CRCs of every zip entry if all of them succeed to be read.
+    *
+    * This method uses `java.util.jar` instead of `java.util.zip` utilities
+    * because the method `getCrc` in [[java.util.zip.ZipEntry]] returns 0 or
+    * -1 values.
+    *
+    * @param jarFile A jar file (which follows the zip specification format).
+    * @return A list of checksums for every zip entry.
+    */
+  def readJarCrcs(jarFile: Path): Option[ArrayBuffer[Long]] = {
```

Comment:
This seems a poor abstraction of a signature
everything is boxed, wrapped with options, just so that we can perform a == on a Seq[Long]

why not def signature(jarFile: Path) : JarSignature

with 2 implementations for CRCs and file attribute based

Reply:
I thought about doing that, but I didn't want to overengineer it. The problem with boxing exists even with your proposed solution.

Reply:
my solution would have no boxing, I am not sure what solution you are referring to as I did not detail any proposal here